lib/Conversion/AffineToStaticLogic/AffineToStaticLogic.cpp

```diff
@@ -0,0 +1,228 @@
+//===- AffineToStaticlogic.cpp --------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#include "circt/Conversion/AffineToStaticLogic.h"
+#include "../PassDetail.h"
+#include "circt/Analysis/SchedulingAnalysis.h"
+#include "circt/Dialect/StaticLogic/StaticLogic.h"
+#include "circt/Scheduling/Algorithms.h"
+#include "circt/Scheduling/Problems.h"
+#include "mlir/Analysis/AffineAnalysis.h"
+#include "mlir/Dialect/Affine/IR/AffineMemoryOpInterfaces.h"
+#include "mlir/Dialect/Affine/IR/AffineOps.h"
+#include "mlir/Dialect/Arithmetic/IR/Arithmetic.h"
+#include "mlir/Dialect/MemRef/IR/MemRef.h"
+#include "mlir/Dialect/StandardOps/IR/Ops.h"
+#include "mlir/IR/ImplicitLocOpBuilder.h"
+#include "mlir/Transforms/LoopUtils.h"
+#include "llvm/ADT/STLExtras.h"
+#include "llvm/Support/Debug.h"
+
+#define DEBUG_TYPE "affine-to-staticlogic"
+
+using namespace mlir;
+using namespace mlir::arith;
+using namespace circt;
+using namespace circt::analysis;
+using namespace circt::scheduling;
+using namespace circt::staticlogic;
+
+namespace {
+
+struct AffineToStaticLogic
+    : public AffineToStaticLogicBase<AffineToStaticLogic> {
+  void runOnFunction() override;
+
+private:
+  void populateOperatorTypes(SmallVectorImpl<AffineForOp> &loopNest);
+  void solveSchedulingProblem(SmallVectorImpl<AffineForOp> &loopNest);
+  void createStaticLogicPipeline(SmallVectorImpl<AffineForOp> &loopNest);
+
+  CyclicSchedulingAnalysis *schedulingAnalysis;
+};
+
+} // namespace
+
+void AffineToStaticLogic::runOnFunction() {
+  // Get scheduling analysis for the whole function.
+  schedulingAnalysis = &getAnalysis<CyclicSchedulingAnalysis>();
+
+  // Collect perfectly nested loops and work on them.
+  auto outerLoops = getOperation().getOps<AffineForOp>();
+  for (auto root : llvm::make_early_inc_range(outerLoops)) {
+    SmallVector<AffineForOp> nestedLoops;
+    getPerfectlyNestedLoops(nestedLoops, root);
+
+    // Restrict to single loops to simplify things for now.
+    if (nestedLoops.size() != 1)
+      continue;
+
+    populateOperatorTypes(nestedLoops);
+    solveSchedulingProblem(nestedLoops);
+    createStaticLogicPipeline(nestedLoops);
```

Comment:
Since at least the first two passes may fail, they could return `LogicalResult` and the `SignalPassFailure` calls around the code could be moved to a single call at the top level.